### PR TITLE
Add an assert wrapper for easy porting.

### PR DESCRIPTION
**Pull request by ezyang**
Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

**Review by apaszke — COMMENTED** [review]
I think it's better to raise C++ exceptions form these asserts. They will be caught in the outermost C function (the one called from Python) and their `.what()` message will be wrapped in a Python error.

**Comment by apaszke** [comment]
You can use a new exception type to make sure no other try blocks catch them (except for `std::exception`)

**Comment by zdevito** [comment]
I am going to merge this now, so we have asserts and consistency checks. We fix it so that primitive ops are easier to write in a latter change if it becomes an issue.

File: torch/csrc/jit/assert.cpp [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  char msg[2048];
+  va_list args;
+
+  va_start(args, fmt);
+  vsnprintf(msg, 2048, fmt, args);
+  va_end(args);
+
+  // Convenience for gdb; helps you avoid fumbling around
+  // to figure out how to trap on throwing a C++ exception.
+  __builtin_trap();
```
Comment (apaszke) [inline torch/csrc/jit/assert.cpp:20]: Just FYI you can stop the process in gdb at every exception if you use `catch throw`. Doesn't thi scause the process to halt immediately, in such a way that the error is never raised?

Comment (ezyang) [reply torch/csrc/jit/assert.cpp:20]: Yes, you are right! I opened a new PR to fix this. https://github.com/ezyang/pytorch/pull/9

**Pull request merged**